Sources/NIOConcurrencyHelpers/fastatomics.swift

```diff
@@ -0,0 +1,269 @@
+}
+
+extension UInt64: FastAtomicPrimitive {
+    public typealias AtomicWrapper = catmc_fast_atomic_unsigned_long_long
+    public static let fast_atomic_create_with_existing_storage = catmc_fast_atomic_unsigned_long_long_create_with_existing_storage
+    public static let fast_atomic_compare_and_exchange         = catmc_fast_atomic_unsigned_long_long_compare_and_exchange
+    public static let fast_atomic_add                          = catmc_fast_atomic_unsigned_long_long_add
+    public static let fast_atomic_sub                          = catmc_fast_atomic_unsigned_long_long_sub
+    public static let fast_atomic_exchange                     = catmc_fast_atomic_unsigned_long_long_exchange
+    public static let fast_atomic_load                         = catmc_fast_atomic_unsigned_long_long_load
+    public static let fast_atomic_store                        = catmc_fast_atomic_unsigned_long_long_store
+}
+
+extension Int: FastAtomicPrimitive {
+    public typealias AtomicWrapper = catmc_fast_atomic_long
+    public static let fast_atomic_create_with_existing_storage = catmc_fast_atomic_long_create_with_existing_storage
+    public static let fast_atomic_compare_and_exchange         = catmc_fast_atomic_long_compare_and_exchange
+    public static let fast_atomic_add                          = catmc_fast_atomic_long_add
+    public static let fast_atomic_sub                          = catmc_fast_atomic_long_sub
+    public static let fast_atomic_exchange                     = catmc_fast_atomic_long_exchange
+    public static let fast_atomic_load                         = catmc_fast_atomic_long_load
+    public static let fast_atomic_store                        = catmc_fast_atomic_long_store
+}
+
+extension UInt: FastAtomicPrimitive {
+    public typealias AtomicWrapper = catmc_fast_atomic_unsigned_long
+    public static let fast_atomic_create_with_existing_storage = catmc_fast_atomic_unsigned_long_create_with_existing_storage
+    public static let fast_atomic_compare_and_exchange         = catmc_fast_atomic_unsigned_long_compare_and_exchange
+    public static let fast_atomic_add                          = catmc_fast_atomic_unsigned_long_add
+    public static let fast_atomic_sub                          = catmc_fast_atomic_unsigned_long_sub
+    public static let fast_atomic_exchange                     = catmc_fast_atomic_unsigned_long_exchange
+    public static let fast_atomic_load                         = catmc_fast_atomic_unsigned_long_load
+    public static let fast_atomic_store                        = catmc_fast_atomic_unsigned_long_store
+}
+
+public class FastAtomic<T: FastAtomicPrimitive> {
+    @usableFromInline
+    typealias Manager = ManagedBufferPointer<Void, T.AtomicWrapper>
+
+    /// Create an atomic object with `value`
+    @inlinable
+    static func makeAtomic(value: T) -> FastAtomic {
+        let manager = Manager(bufferClass: self, minimumCapacity: 1) { _, _ in }
```

Comment:
@atrick the only reason we can use an `UInt8` buffer of the right size is that we need memory that’s sufficiently aligned for an atomic. I assume there are no alignment guarantees for this?

Reply:
@weissi I'm not sure what you're asking. `ManagedBuffer` gives you the alignment guarantee per its `_alignmentMask`. `Unsafe[Buffer]Pointer<T>.allocate` would give you the same guarantee. Or you can explicitly provide alignment to `UnsafeRawPointer.allocate`

Reply:
I'm a bit confused by @weissi's question here: we're not allocating a `UInt8` buffer, we're allocating a buffer of type `T.AtomicWrapper` of size 1. This does not seem like it would have any alignment issues at all.

Reply:
@atrick / @Lukasa sorry, I was on a phone only on Friday. What we really want to express here is tail allocation for an opaque (forward declared) struct in C. Of course because it's opaque, we don't know it's true length/alignment requirements. We could however expose getters for length & alignment.

That way, we could make a `ManagedBufferPointer` that holds `length` `UInt8`s. Then we could pass the `UnsafeMutableRawBufferPointer` to these `UInt8`s to the C function that initialises the atomic. The only issue here is that our `UnsafeMutableRawBufferPointer` isn't correctly aligned to fit the struct that contains the atomic.

And I was wondering if we could ask `ManagedBufferPointer` to give us tail allocated storage aligned to a user-chosen alignment. So something like

    let manager = Manager<Void, UInt8>(bufferClass: self, minimumCapacity: get_width_for_opaque_struct(), alignment: get_alignment_requirement_for_opaque_struct()) { _, _, in }

But this API doesn't seem to exist. We could obviously ask for a possibly larger size and align ourselves.

Reply:
@weissi Ah. I get it. I think it would be reasonable to propose adding an (over)alignment option to ManagedBuffer.create. I'm not sure if we can add an underscored, defaulted argument to a public API. Might need to add a new underscored API instead. Without an API I think you need to allocate extra space.